src/term/string.rs

```diff
@@ -0,0 +1,399 @@
+                f,
+                "index out of bounds. 
+                 Expected 2nd argument (start) to be between 0 and {str_len}, got {start}"
+            ),
+            EndOutOfBounds {
+                start,
+                end,
+                str_len,
+            } => write!(
+                f,
+                "index out of bounds. 
+                Expected 3rd argument (end) to be between {start} and {str_len}, got {end}"
+            ),
+        }
+    }
+}
+
+/// Types and functions designed to make it as easy as possible not to accidentally
+/// break up Unicode extended grapheme clusters in string operations.
+mod grapheme_cluster_preservation {
+    use unicode_segmentation::GraphemeCursor;
+
+    /// Returns a [`SearchIter`], which is an `Iterator` for iterating through
+    /// `haystack` in order to find `needle`.
+    ///
+    /// [`SearchIter`] produces [`SearchEvent`]s which can be used to
+    /// implement a variety of different behaviours.
+    pub fn search<'a>(src: &'a str, needle: &'a str) -> SearchIter<'a> {
+        SearchIter {
+            haystack: src,
+            needle,
+            cursor: GraphemeCursor::new(0, src.len(), true),
+            current_split_start: 0,
+            last_boundary: 0,
+        }
+    }
+
+    /// An event emitted by [`SearchIter`].
+    pub enum SearchEvent<'a> {
+        /// The `needle` was found within `haystack`.
+        /// Since the caller already has a reference to `needle`, we don't
+        /// include it in this event. However, we do include every part of
+        /// `haystack` which was traversed up until the match happened.
+        Match {
+            /// The slice of the string from the end of the last match
+            /// up to this one.
+            since_last_match: &'a str,
+        },
+        /// Iteration through `haystack` has finished, but there was
+        /// an unconsumed (non-matching) slice which the caller may wish
+        /// to do something with.
+        LastNonMatch {
+            /// Any non-matching text from the end of the string.
+            non_match: &'a str,
+        },
+    }
+
+    /// Our base algorithm for implementing a variety of grapheme cluster
+    /// preserving methods on [`NclString`]. As the name suggests, it is an
+    /// `Iterator`. However, rather than iterating through grapheme clusters
+    /// directly, its `Item` is a [`SearchEvent`].
+    pub struct SearchIter<'a> {
+        haystack: &'a str,
+        needle: &'a str,
+        cursor: GraphemeCursor,
+        current_split_start: usize,
+        last_boundary: usize,
+    }
+
+    impl<'a> Iterator for SearchIter<'a> {
+        type Item = SearchEvent<'a>;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            let unwrap_explanation = "
+                None of the GraphemeIncomplete errors are possible here:
+                    - PreContext and PrevChunk only happen if chunk_start is nonzero.
```

Comment:
same thing here, I don't think this will give the indentation you expect. However with indented text inside indented text you're out of luck: either you don't use multiline string but you need to indent everything on the left since the very first column, or...I don't now :man_shrugging: use a macro like `indent_doc!` provided by the eponymous crate, I believe